Src/main.c

```diff
@@ -100,14 +103,41 @@ int main(void)
   debug("\n\nBootcamp starting up...");
   debug("Compiled on %s at %s", __DATE__, __TIME__);
 
-
   HAL_GPIO_WritePin(LED1_GPIO_Port, LED1_Pin, GPIO_PIN_SET);
+
+  HAL_ADC_Start(&hadc); //Enables ADC and starts conversion of the regular channels.
+  HAL_TIM_PWM_Start(&htim16, TIM_CHANNEL_1); //starts PWM signal generation. First argument is the pointer to the TIM handle and the second parameter is the channel used or the timer
+
+
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
+  uint16_t ADCOutput = 0; // stores converted value from data register
+  float PWMCompareDecimal = 0.0;
+  uint16_t PWMCompare = 0; //we only need to provide a 16 bit input for the pwm
+
+  // sets peripheral values (constants)
+  const int MINIMUM_PWM_VALUE = 3000; //minimum value of the duty cycle (MIN_VALUE / 60000 * TIM16.Period)
+  const int PWM_COMPARE_VALUE_RANGE = 3000; //Used to determine the range of the duty cycle (RANGE / 60000 * TIM16.Period)
+  const int ADC_MAX_VALUE = 4095; // Max adc output possible 
+
   while (1)
   {
+    ADCOutput = HAL_ADC_GetValue(&hadc); //Gets the converted value from data register of regular channel. (A value between 0 and 4095 representinng 0 to 3.3 V [assuming a 12-bit ADC is being used])
+    /*
+    
+      From my understanding, the __HAL_TIM_SET_COMPARE() function compares the pwm_compare value to the period of the TIM16. The value a = pwm_compare/TIM16.Period*100% is the duty period of the PWM.
+      Since we want 1ms to 2ms on-time, we are looking at a 5-10% duty cycle. Right now, the PWM period is 60000, thus we want the pwm_compare value to have a range of 3000 to 6000!
+
+    */
+
+    // Converts value to a range of 3000 (1 ms on-time) to 6000 (2 ms on-time) 
+    PWMCompareDecimal = round((float) ADCOutput * (float) PWM_COMPARE_VALUE_RANGE / (float) ADC_MAX_VALUE + (float) MINIMUM_PWM_VALUE); //calculates using float first so truncation doesn't result in the PWMCompare value being one less than it should be. 
```

Comment:
Curious, if you round off the input here, why does PWMCompareDecimal need to be a float vs an int? I might be missing something but if my understanding of the code is correct, you shouldn't have decimal places here, and we can save ourselves a few extra bytes by using a more appropriate data type if there is one available. As I mentioned, memory ain't free :)

Reply:
My bad. I was under the impression round() returned a float/double 😬 

... we could always download more memory... 👀  XD